GovAlta/ui-components
# feature/DDIDS-1111-vscode-extension: Adding the VS Code extension sup…

### Description
When teams are consuming the web components in an Angular project, they will not get IntelliSense or Code Snippets when writing code. This makes understanding available properties much more difficult for Angular developers using our web components. 

### How to test
1. In the code (of this repo), run `npm run build:all` or `npm run build:web-components`
2. Under the build folder `dist/libs/web-components/` you will see a new generated file `html.html-data.json`
3. Use VS Code and open this repo, now come to`apps/angular-demo/src/app/accordion` for example, under `accordion.component.html`, hover over the `goa-accordion` element, you will see the pop-up description that is generated by VS code. It doesn't exist. 

If you clone the repo from [Angular Template](https://github.com/GovAlta/ui-components-angular-template), then follow the steps below to test: 
1. Clone this repo, run `npm run build:all`
2. Under Angular repo that was created by Angular template, under package.json, change the value `"@abgov/web-components": "/Users/thytran142/code/ui-components/dist/libs/web-components",` (the directory of your ui-components local)
3. Under your Angular repo, run `npm install` to update your node_modules
4. Under your `.vscode/settings.json`, add the below code:
```
{
    "html.customData": [
      "./node_modules/@abgov/web-components/html.html-data.json"
    ]
  }
```
5. Quit VS code and open the Angular project again, hover an element or add an element, Intellisense feature is ON now. 
![image](https://user-images.githubusercontent.com/120135417/219513355-ea32f4c0-92a0-4bdb-94bf-1f3b85f9509d.png)

### Stories
[DDIDS-1111](https://goa-dio.atlassian.net/jira/software/c/projects/DDIDS/boards/185?modal=detail&selectedIssue=DDIDS-1111)

### What changed
I tried to use the `cem-plugin-vs-code-custom-data-generator` and `js doc` to automatically generate web-component.html-data.json file, but **it doesn't work for svelte components.** 
![image](https://user-images.githubusercontent.com/120135417/219510002-4da2f763-0f0f-4f4e-a0f1-723e2c7b50da.png)

**Solution:** 
The hover function under VSCode: 
This is supported through the `html.customData` setting found in [`.vscode/settings.json`](.vscode/settings.json):
 
```json
{
  "html.customData": [
    "./dist/libs/web-components/html.html-data.json"
  ],
}
```
* So under each folder `web-components/components/*`, we can document the description for our web component with *.html-data.json file. Format is like the below:
```
{
      "name": "my-button",
      "description": "My button. You should use it as in `<my-button type='alert'></mybutton>`.",
      "attributes": [
        {
          "name": "type",
          "description": "My button type",
          "values": [{ "name": "alert" }]
        }
      ],
      "references": [
        {
          "name": "Bootstrap buttons",
          "url": "https://getbootstrap.com/docs/4.0/components/buttons/"
        }
      ]
    }
```

* Under the package.json, I add the below script: 
"postbuild:web-components": "node libs/web-components/custom-element-manifest-analyze.js",

The custom-element-manifest-analyze.js will collect all *.html-data.json, sum up them and put `html.html-data.json` to the build folder. 

### Screenshot
![image](https://user-images.githubusercontent.com/120135417/219512086-4cafc726-3cf8-46de-b6a7-2b7d09153f71.png)
